crates/astria-bridge-contracts/src/lib.rs

```diff
@@ -1,4 +1,654 @@
+    CallBaseChainAssetPrecision {
+        source: Box<dyn std::error::Error + Send + Sync + 'static>,
+    },
+    #[error("ics20 asset must have a channel to be withdrawn via IBC")]
+    Ics20AssetWithoutChannel,
+    #[error("could not parse ics20 asset channel as channel ID")]
+    ParseIcs20AssetSourceChannel { source: ibc_types::IdentifierError },
+}
+
+pub struct NoProvider;
+pub struct WithProvider<P>(Arc<P>);
+
+pub struct GetWithdrawalActionsBuilder<TProvider = NoProvider> {
+    provider: TProvider,
+    contract_address: Option<ethers::types::Address>,
+    bridge_address: Option<Address>,
+    fee_asset: Option<asset::Denom>,
+    sequencer_asset_to_withdraw: Option<asset::Denom>,
+    ics20_asset_to_withdraw: Option<asset::TracePrefixed>,
+}
+
+impl Default for GetWithdrawalActionsBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl GetWithdrawalActionsBuilder {
+    #[must_use]
+    pub fn new() -> Self {
+        Self {
+            provider: NoProvider,
+            contract_address: None,
+            bridge_address: None,
+            fee_asset: None,
+            sequencer_asset_to_withdraw: None,
+            ics20_asset_to_withdraw: None,
+        }
+    }
+}
+
+impl<P> GetWithdrawalActionsBuilder<P> {
+    #[must_use]
+    pub fn provider<Q>(self, provider: Arc<Q>) -> GetWithdrawalActionsBuilder<WithProvider<Q>> {
+        let Self {
+            contract_address,
+            bridge_address,
+            fee_asset,
+            sequencer_asset_to_withdraw,
+            ics20_asset_to_withdraw,
+            ..
+        } = self;
+        GetWithdrawalActionsBuilder {
+            provider: WithProvider(provider),
+            contract_address,
+            bridge_address,
+            fee_asset,
+            sequencer_asset_to_withdraw,
+            ics20_asset_to_withdraw,
+        }
+    }
+
+    #[must_use]
+    pub fn contract_address(self, contract_address: ethers::types::Address) -> Self {
+        Self {
+            contract_address: Some(contract_address),
+            ..self
+        }
+    }
+
+    #[must_use]
+    pub fn bridge_address(self, bridge_address: Address) -> Self {
+        Self {
+            bridge_address: Some(bridge_address),
+            ..self
+        }
+    }
+
+    #[must_use]
+    pub fn fee_asset(self, fee_asset: asset::Denom) -> Self {
+        Self {
+            fee_asset: Some(fee_asset),
+            ..self
+        }
+    }
+
+    #[must_use]
+    pub fn sequencer_asset_to_withdraw(self, sequencer_asset_to_withdraw: asset::Denom) -> Self {
+        self.set_sequencer_asset_to_withdraw(Some(sequencer_asset_to_withdraw))
+    }
+
+    #[must_use]
+    pub fn set_sequencer_asset_to_withdraw(
+        self,
+        sequencer_asset_to_withdraw: Option<asset::Denom>,
+    ) -> Self {
+        Self {
+            sequencer_asset_to_withdraw,
+            ..self
+        }
+    }
```

Comment:
why have both of these?

Reply:
Nice to maintain a builder chain. Could only have the `set_*` one, but I like to maintain synchrony with the other setters.